### PR TITLE
Issue #11292: Nested Boolean Compares

### LINKED ISSUE #11292: Comparisons of structs with NULL fields turn out wrong
### What happens?

The comparison of two struct values will return a wrong (non-`NULL`) result if relevant fields contain `NULL`.  The behaviour of DuckDB deviates from that of SQLite3 and PostgreSQL.

EDIT: I am aware of the discussion of _Comparison Operators_ on https://duckdb.org/docs/sql/data_types/struct.  I am skeptical of DuckDB's current rationale that the behaviour of the comparison with `NULL` is dependent on whether `NULL` appears "top-level" or in a struct field.  That violates the principles of least surprise and compositionality.

### To Reproduce

# DuckDB

OK (field `b` is not relevant to decide the comparison):
```
> SELECT {a:1, b:2} < {a:2, b:NULL} AS OK;
┌─────────┐
│   OK    │
│ boolean │
├─────────┤
│ true    │
└─────────┘
```

Bug (field `b` is relevant, the comparison of `2` with `NULL` should yield `NULL` and the entire struct comparison is thus expected to return `NULL`; in other words: embedding `NULL` into a struct should not render `NULL` comparable):
```
> SELECT {a:1, b:2} < {a:1, b:NULL} AS Bug;
┌─────────┐
│   Bug   │
│ boolean │
├─────────┤
│ true    │
└─────────┘
```

Likewise:

```
> SELECT (1,1) < (2,NULL) AS OK;
┌─────────┐
│   OK    │
│ boolean │
├─────────┤
│ true    │
└─────────┘
> SELECT (1,1) < (1,NULL) AS Bug;
┌─────────┐
│   Bug   │
│ boolean │
├─────────┤
│ true    │
└─────────┘
```

# PostgreSQL (v15.6)

```
> SELECT (1,1) < (2,NULL) AS OK;
┌────┐
│ ok │
├────┤
│ t  │
└────┘

> SELECT (1,1) < (1,NULL) AS OK;
┌──────┐
│ ok   │
├──────┤
│ NULL │
└──────┘
```

# SQLite3 (v3.39.5)

```
> SELECT (1,1) < (2,NULL);
1

> SELECT ((1,1) < (1,NULL)) IS NULL;
1
```

### OS:

macOS

### DuckDB Version:

v0.10.1-dev397 d745e29fd1

### DuckDB Client:

CLI

### Full Name:

Torsteh Grust

### Affiliation:

U Tübingen, Germany

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by hawkfish**
We used to have somewhat strange comparison semantics for nested types
where the top level produced NULL when comparing NULLs but we used
DISTINCT semantics lower down. This was surprising, didn't match Postgres,
and wasn't stable if you pulled the contents up a level e.g., via UNNEST.

Several tests were relying on these semantics. Most of them could
be fixed by changing them to using DISTINCT comparisons, but a small number
were producing incorrect results and have been corrected.

fixes: https://github.com/duckdb/duckdb/issues/11292
fixes: https://github.com/duckdblabs/duckdb-internal/issues/1657

### CODE REVIEW

File: src/include/duckdb/common/vector_operations/vector_operations.hpp [hunk 1]
```diff
@@ -87,17 +87,19 @@ struct VectorOperations {
 	// Select Comparisons
 	//===--------------------------------------------------------------------===//
 	static idx_t Equals(Vector &left, Vector &right, const SelectionVector *sel, idx_t count, SelectionVector *true_sel,
-	                    SelectionVector *false_sel);
+	                    SelectionVector *false_sel, ValidityMask *null_mask = nullptr);
```
Comment (lnkuiper) [inline src/include/duckdb/common/vector_operations/vector_operations.hpp:90]: Could these pointers to selection vectors be `optional_ptr`s instead?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Thanks for fixing this! I just have one remark:

File: src/include/duckdb/common/vector_operations/vector_operations.hpp (hunk 1 above)
Comment (hawkfish) [reply src/include/duckdb/common/vector_operations/vector_operations.hpp:90]: They could - I was just matching the existing style. Do we change all pointers, just the new ones, or leave it for another cleanup pass?

Comment (hawkfish) [reply src/include/duckdb/common/vector_operations/vector_operations.hpp:90]: I'll start with the new argument then. My concern about the existing arguments was just a concern that I would have to change all the callers.

**Comment by hawkfish** [comment]
This changes the results of nested comparisons as Boolean functions (but not straight comparisons). This might break someone's queries and should appear in the breaking changes list once it is merged. But its a bit obscure, so down the bottom of the list is fine.

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes! LGTM

**Comment by Mytherin** [comment]
Thanks!

**Comment by Teggy** [comment]
Wonderful! Thank you so much, folks!

**Pull request merged**